### PR TITLE
NAPI JSC updates

### PR DESCRIPTION
**Pull request by bghgary**
Fixes shutdown issues in BabylonReactNative repo.

- Retain JSC global context.
- Track references and delete them on env detach.
- Implement weak references properly.

### CODE REVIEW

File: Dependencies/napi/source/js_native_api_JavaScriptCore.cc [hunk 1]
```diff
@@ -205,7 +202,8 @@ namespace {
       JSObjectSetPrototype(env->context, constructor, prototype);
       
       JSValueRef exception{};
-      JSObjectSetProperty(env->context, prototype, JSString("constructor"), constructor, kJSPropertyAttributeReadOnly | kJSPropertyAttributeDontDelete, &exception);
+      JSObjectSetProperty(env->context, prototype, JSString("constructor"), constructor,
```
Comment (ryantrem) [inline Dependencies/napi/source/js_native_api_JavaScriptCore.cc:205]: Nit: Is this line break intentional? I thought it was perfectly readable on one line, so wasn't sure if this was some automatic tooling.

File: Dependencies/napi/source/js_native_api_JavaScriptCore.cc [hunk 2]
```diff
@@ -464,6 +505,86 @@ namespace {
...
+    , _count{count} {
+  }
+  
+  napi_status init(napi_env env) {
+    // track the ref values to support weak refs
+    auto pair{env->active_ref_values.insert(_value)};
+    if (pair.second) {
+      ExternalInfo* info{};
+      CHECK_NAPI(ExternalInfo::Wrap(env, _value, &info));
+      info->AddFinalizer([iter{pair.first}](ExternalInfo* info) {
```
Comment (ryantrem) [inline Dependencies/napi/source/js_native_api_JavaScriptCore.cc:520]: So an interator for an `unordered_set` remains valid when the collection is mutated?

File: Plugins/NativeEngine/Source/NativeEngine.cpp [hunk 3]
```diff
@@ -467,10 +467,16 @@ namespace Babylon
...
     {
         auto callback = info[0].As<Napi::Function>();
-        arcana::make_task(m_runtimeScheduler, m_cancelSource, [this, callbackRef = Napi::Persistent(callback)]() {
+
+        if (m_requestAnimationFrameCalback.IsEmpty() ||
+            m_requestAnimationFrameCalback.Value() != callback) {
+            m_requestAnimationFrameCalback = Napi::Persistent(callback);
+        }
+
+        arcana::make_task(m_runtimeScheduler, m_cancelSource, [this]() {
```
Comment (ryantrem) [inline Plugins/NativeEngine/Source/NativeEngine.cpp:476]: Use Dispatch directly.

File: Dependencies/napi/source/js_native_api_JavaScriptCore.cc (hunk 1 above)
Comment (bghgary) [reply Dependencies/napi/source/js_native_api_JavaScriptCore.cc:205]: I did this manually in this case. I try to keep NAPI conventions which seems to limit the line length.

File: Dependencies/napi/source/js_native_api_JavaScriptCore.cc (hunk 2 above)
Comment (bghgary) [reply Dependencies/napi/source/js_native_api_JavaScriptCore.cc:520]: Good catch. I was originally using a different container that was okay, but unordered_set is not okay.

https://en.cppreference.com/w/cpp/container/unordered_set
> Operations | Invalidated
> -- | --
> All read only operations, swap, std::swap | Never
> clear, rehash, reserve, operator= | Always
> insert, emplace, emplace_hint | Only if causes rehash
> erase | Only to the element erased
> 
>

**Pull request merged**